sympy/matrices/random.py

```diff
@@ -0,0 +1,2378 @@
+
+    Explanation
+    ===========
+    Let $\mathbf{I}$ be the $n \times n$ identity matrix
+    and $\mathbf{D}$ an diagonal matrix with only zero and one entries
+    of rank $r$ (i.e. all entries but $n-r$ diagonal entries are zero).
+
+    If a full rank matrix $\mathbf{A}$ is given, then
+
+    .. math::
+
+        \mathbf{B}
+        = \mathbf{D} \cdot
+        (\mathbf{I} + \mathbf{A} \cdot (\mathbf{I}-\mathbf{D}))
+        = \mathbf{D} + \mathbf{D} \cdot \mathbf{A} \cdot(\mathbf{I}-\mathbf{D})
+
+    is of rank $r$, too.  Some for $\mathbf{A}\cdot\mathbf{B}$
+    which contains $r$ columns with $\mathbf{A}$.
+
+    If $\mathbf{A}$ is an upper triangular matrix $\mathbf{B}$
+    as well $\mathbf{A} \cdot \mathbf{B}$ is one, too.
+
+    Here, $\mathbf{A} \cdot \mathbf{B}$ is returned.
+    Note, if $n=r$ the matrix $\mathbf{B}$ is the identity.
+
+    Examples
+    ========
+    >>> from sympy import Matrix
+    >>> from sympy.matrices.random import regular_to_singular
+    >>> m = Matrix([[1, 2, 3],[4, 5, 6], [7, 8 ,0]])
+    >>> m.rank()
+    3
+    >>> n = regular_to_singular(m, 2)
+    >>> n
+    Matrix([
+    [1, 2, 15],
+    [4, 5, 42],
+    [7, 8, 69]])
+    >>> n.rank()
+    2
+
+    Parameters
+    ==========
+    mat : Matrix
+        a complex matrix
+    rank : integer
+        rank of matrix
+
+    """
+    if rank is None:
+        return mat
+    dim = _shape(mat)[0]
+    if rank == dim:
+        return mat
+    i = _eye(dim)
+    p = permutation(dim)
+    d = projection(dim, (0, rank))
+    d = _multiply(p.inv(), d, p)
+    md = _multiply(mat, d)
+    return md + _multiply(md, mat, i - d)
+
+
+# === base matrices ===
+
+
+def projection(dim,
+               index=None):
+    r"""A randomly generated n x n projection matrix
+
+    Explanation
+    ===========
+
+    A projection is a identity like matrix
+    but with zero diagonal entiries off *index*.
+
+    Examples
+    ========
+
+    >>> from sympy.core.random import seed
+    >>> from sympy.matrices.random import projection
+    >>> seed(1)
+
+    >>> projection(3)
+    Matrix([
+    [0, 0, 0],
+    [0, 1, 0],
+    [0, 0, 0]])
+
+    >>> projection(3, index=(1,3))  # no more random
+    Matrix([
+    [0, 0, 0],
+    [0, 1, 0],
+    [0, 0, 1]])
+
+    Parameters
+    ==========
+
+    dim : integer
+        dimension n of matrix
+    index : tuple(integer, integer) (optional)
+        coordinates ``(i,j)`` for start (included) and end (excluded)
+        to project onto
+
+    See Also
+    ========
+    jordan
+    elementary
+    diagonal_normal
+
+    """
+
+    index = index or _sample(range(dim + 1), 2)
+    obj = _eye(dim)
+    start, end = sorted(index)
+    for i in range(dim):
+        v = 1 if start <= i < end else 0
+        obj[i, i] = v
+    return obj
+
+
+def jordan(dim,
+           index=None,
+           scalar=None):
+    r"""n x n matrix with a single Jordan block of given eigenvalue
```

Comment:
I find this function slightly odd. It returns an identity matrix with a given Jordan block. It would make more sense to me to have a function that takes a list of `(eigenvalue, blocksize)` pairs e.g.:
```
>>> jordan([(2, 2), (3, 1)])
⎡2  1  0⎤
⎢       ⎥
⎢0  2  0⎥
⎢       ⎥
⎣0  0  3⎦
```
The only random element here seems to be choosing the eigenvalue for the nontrivial Jordan block. Maybe that could be done by only specifying the sizes of the blocks e.g. `jordan([2, 1])`.

Reply:
As the module has the concept of _basis types matrices_ which usually operate on a single subspace of a vector space (hence have a single non trivial block). Here `jordan` is such basic type and acts on a single [space of generalized eigenvectors](https://en.wikipedia.org/wiki/Generalized_eigenvector).

In order to keep the signature consistent for all basic matrices, `index` sets start and end indices of the non trivial block.

Multiple Jordan blocks come into play with the `jordan_normal` matrix which are build by blocks of _basis type matrices_. Same as `isometry_normal` is build by multiple `rotation` blocks. The `spec` argument of _normal form matrices_ follows exactly the pattern `(eigenvalue, blocksize)` you suggest.